engine/component.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <iostream>
+#include <memory>
+#include <vector>
+
+namespace pong
+{
+
+class GameObject;
+
+template <typename BaseClass>
+class TypeIDGenerator
+{
+public:
+    template<class SubClass>
+    static const int GetID(bool newInstance = false)
+    {
+        static int instanceCount = 0;
+        static const int idCounter = sID++;
+
+        return idCounter + (newInstance ? instanceCount++ : 0);
+    }
+
+private:
+    static int sID;
+};
+
+template<typename T> int TypeIDGenerator<T>::sID = 0;
+
+template<typename ComponentSubType>
+int GetComponentTypeID()
+{
+    return TypeIDGenerator<BaseComponent>::GetID<ComponentSubType>();
+}
+
+template<class T>
+class Component : public BaseComponent
+{
+public:
+    virtual ~Component() = default;
+
+    static void AddComponent(std::unique_ptr<T> component)
+    {
+        mComponents.push_back(std::move(component));
+    }
+
+    static const std::vector<std::unique_ptr<T>>& GetComponents()
+    {
+        return mComponents;
+    }
+
+    static std::vector<std::unique_ptr<T>> mComponents;
```

Comment:
Is there any other way to do this sort of thing? Essentially, with an unknown number of component types, it needs to create a vector for each component type even if it is user defined.

Reply:
No, I think this is fine I think. I can't think of another way to do it.